MediaGallerySynchronization/Model/MediaStorageFilesBatchGenerator.php

```diff
@@ -0,0 +1,129 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\MediaGallerySynchronization\Model;
+
+use Magento\Framework\Filesystem;
+use Magento\Framework\App\Filesystem\DirectoryList;
+use Magento\Framework\Filesystem\Directory\Read;
+use Magento\MediaGalleryApi\Api\IsPathBlacklistedInterface;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Media storage assets batch generator
+ */
+class MediaStorageFilesBatchGenerator
+{
+    private const IMAGE_FILE_NAME_PATTERN = '#\.(jpg|jpeg|gif|png)$# i';
+    
+    /**
+     * @var GetAssetsIterator
+     */
+    private $getAssetsIterator;
+
+    /**
+     * @var Filesystem
+     */
+    private $filesystem;
+
+    /**
+     * @var Read
+     */
+    private $mediaDirectory;
+
+    /**
+     * @var IsPathBlacklistedInterface
+     */
+    private $isPathBlacklisted;
+
+    /**
+     * @var array
+     */
+    private $assetsPaths;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $log;
+    
+    /**
+     * @param LoggerInterface $log
+     * @param IsPathBlacklistedInterface $isPathBlacklisted
+     * @param Filesystem $filesystem
+     * @param GetAssetsIterator $assetsIterator
+     */
+    public function __construct(
+        LoggerInterface $log,
+        IsPathBlacklistedInterface $isPathBlacklisted,
+        Filesystem $filesystem,
+        GetAssetsIterator $assetsIterator
+    ) {
+        $this->log = $log;
+        $this->isPathBlacklisted = $isPathBlacklisted;
+        $this->getAssetsIterator = $assetsIterator;
+        $this->filesystem = $filesystem;
+        $this->assetsPaths = [];
+    }
+
+    /**
+     * Return files from files system by provided size of batch
+     *
+     * @param int $size
+     */
+    public function getFiles(int $size): \Generator
```

Comment:
I think the size should be moved to di configuration of this class for:
 - consistency with database batch generator and easier utilization
 - this class has a very specific purpose, so it's unlikely a different batch size would be required for different implementations
```suggestion
    public function execute(): \Traversable
```